### ISSUE TITLE
Strange write speed

### ISSUE THREAD

**Issue by guestisp** [lizardfs/lizardfs#599]
Something strange is happening here.

goal=3, 3 chunkservers with 2 gigabit nic bonded together (balance-rr). 
"iperf" show about 2gbit in a single connection (thanks to balance-rr bonding mode)

Writing from a client (with the same bonding mode) i'm able to get about 180MB/s (that's impressive) but writing directly on the chunkserver i'm getting only about 90MB/s:

From LizardFS mountpoint:
```
# dd if=/dev/zero of=test bs=1M count=1000 oflag=dsync
1000+0 records in
1000+0 records out
1048576000 bytes (1.0 GB, 1000 MiB) copied, 5.75394 s, 182 MB/s
```

Directly from chunkserver, totally skipping fuse and lizard:
```
dd if=/dev/zero of=test bs=1M count=1000 oflag=dsync
1000+0 record dentro
1000+0 record fuori
1048576000 byte (1,0 GB) copiati, 11,392 s, 92,0 MB/s
```

That's non-sense, lizard can't be faster than a single disk, when writing in sync.
Disks should be the bottleneck, as you have to wait for the ACK from each disk.
It seems that fsync is ignored or something strange is happening.
When writing through Lizard i'm able to reach about 1550mbps by using 2 gigabit nics.

`PERFORM_FSYNC` is as default (1) on all chunkservers

```
# mount | grep mfs
mfs#mfsmaster:9421 on /mnt type fuse (rw,nosuid,nodev,noexec,noatime,user_id=0,group_id=0,default_permissions,allow_other)
```

**Comment by Blackpaw** [comment]
What are your full mount options?

`ps -aux|grep mfsmount`

**Comment by guestisp** [comment]
`mfsmount -o big_writes,nosuid,node
v,noatime,noexec /mnt/`

**Comment by DarkHaze** [comment]
Maybe the file/directory you test have different goal than you expect. This
bahaviour would be typical for xor or ec goal.

01.10.2017 14:27 "GUEST.it" <notifications@github.com> napisał(a):

**Comment by guestisp** [comment]
I'm only using goal=3
But even with goal=1 or anything else, lizard can't be faster than the underlying disks, if writes are all synchronous

**Comment by psarna** [comment]
Well, it can. Local `dd` with dsync flag is not the same operation as writing to fuse with direct I/O. When chunkservers write data, it's not equal to doing dd with dsync, but to writing certain blocks and syncing it afterwards. It's a totally different syscall pattern, can possibly sync with hardware at totally different times, and so it can be faster than `dd` with dsync. The best way of checking it is to run some kind of strace/perf on chunkservers and see if they actually sync data with fsync or flush. Some tools like iotop can be of help as well.

**Comment by guestisp** [comment]
Uhm, but why writing from the host to the mfs mount point i'm getting about 180MB/s and writing from a VM booted from lizardfs i'm still capped at 30 MB/s ?

**Comment by guestisp** [comment]
So, if the writer wants to be sure that data are really wrote on disks, it can't because even by using dsync data is not always wrote to disk in sync.

**Comment by psarna** [comment]
With PERFORM_FSYNC=1 it is, data is written an synced before write operation is confirmed. Still, writing to a distributed system through fuse follows different patterns than dd, so speed of these operations may be different.

Without checking it via some profiling tools (and more info on hardware, OS kernel parameters, etc) I can't guess what is the reason for all these differences.

**Comment by guestisp** [comment]
Ok but regardless the writing pattern, if data is written an synced before write operation is confirmed, disks speed will always be the bottleneck.

You can't reach 180MB/s if *any* disks is reaching 100MB/s at maximum, regardless the writing pattern.
If you are faster than disk, obviously you have some "cache" somewhere, because disk can't ack to any write faster than it's speed. If disks are 100MB/s, ack would be "sent" at 100MB/s, thus 180MB/s are impossible.

**Comment by psarna** [comment]
True. Still, profiling and strace of both dd and chunkservers will tell what exactly happens. I never encountered a situation where LizardFS showed speeds higher than chunkserver disk hardware limit, so I don't have an answer ready.

**Comment by psarna** [comment]
*with regular goals and sync, otherwise it's common to be faster with caches and ec

**Comment by guestisp** [comment]
Today or tomorrow I'll try to strace the chunkserver process during a "dd"

And what about the huge performance drop seen inside a VM booted from lizardfs?
I can understand a little overhead, but here I'm seeing a 6x drop. 180MB/s to 30MB/s

**Comment by psarna** [comment]
That depends on so many things, there's no easy answer. My private virtual machines made for testing purposes run on LizardFS and don't have such a performance overhead. Even things like block size of a file system inside the machine may impact performance a lot, not mentioning other file system and kernel parameters.

**Comment by guestisp** [comment]
Do you need a full strace dump or just some specific calls (if yes, which calls?)

**Comment by guestisp** [comment]
This is a strace on chunkserver process when writing 1GB file with `dd if=/dev/zero of=prova bs=1M count=1000 oflag=direct`

```
[pid 12726] open("/export/sda/chunk/chunks00/chunk_00000000000015A2_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 11
[pid 12725] open("/export/sda/chunk/chunks00/chunk_00000000000015A3_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 21
[pid 12724] open("/export/sda/chunk/chunks00/chunk_00000000000015A4_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 22
[pid 12732] open("/export/sda/chunk/chunks00/chunk_00000000000015A5_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 23
[pid 12731] open("/export/sda/chunk/chunks00/chunk_00000000000015A6_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 19
[pid 12733] open("/export/sda/chunk/chunks00/chunk_00000000000015A7_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 24
[pid 12730] open("/export/sda/chunk/chunks00/chunk_00000000000015A8_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 25
[pid 12729] open("/export/sda/chunk/chunks00/chunk_00000000000015A9_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 11
[pid 12728] open("/export/sda/chunk/chunks00/chunk_00000000000015AA_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 21
[pid 12727] open("/export/sda/chunk/chunks00/chunk_00000000000015AB_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 26
[pid 12726] open("/export/sda/chunk/chunks00/chunk_00000000000015AC_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 22
[pid 12725] open("/export/sda/chunk/chunks00/chunk_00000000000015AD_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 23
[pid 12724] open("/export/sda/chunk/chunks00/chunk_00000000000015AE_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 19
[pid 12732] open("/export/sda/chunk/chunks00/chunk_00000000000015AF_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 24
[pid 12731] open("/export/sda/chunk/chunks00/chunk_00000000000015B0_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 25
[pid 12733] open("/export/sda/chunk/chunks00/chunk_00000000000015B1_00000001.mfs", O_RDWR|O_CREAT|O_TRUNC, 0666) = 11
```

as you can't see, there isn't any `O_DIRECT` or `O_SYNC` flag passed when opening the file un the underlying disk. Thus, writes are always asynchronous regardless what client is asking for.

**Comment by guestisp** [comment]
`fsync` is called *after* chunk is wrote. That's not the same as honoring what client is asking for, in example, writing synchronous and could lead to some (bad) race condition.

If you loose a chunk server between a `write` and it's subsequent `fsync` you'll loose data.
By opening a file with `O_DIRECT` you can't loose data because write is wrote directly to disks bypassing the cache.

**Comment by psarna** [comment]
If you lose a chunkserver in the middle of writing, then this write operation would not be acked by master and will return error to client instead. The same thing would happen if your disk broke in the middle of writing.

O_DIRECT/O_SYNC would not help you in this case, it will most likely make things worse - you'd just end up with partially written chunk and client would still receive an error - because the operation couldn't be completed ("no chunkservers available to finish the operation"). With O_DIRECT, restarting a chunkserver in the middle of writing would basically guarantee that this chunk becomes inconsistent. Without it, there's a fair chance that data inside is still consistent.

With PERFORM_FSYNC on, it goes as follows:
client sends data -> chunkserver receives data -> chunkserver writes data to disk -> chunkservers syncs data so it's flushed from cache to hardware -> client receives information that operation succeeded

This is synchronous - client waits until the operation fully succeeds and all data is written to disk.

Also, what race condition? I can't see any. Each chunkserver locks a chunk before it starts any I/O operations on it, so there's nobody to race with.

**Comment by guestisp** [comment]
> With PERFORM_FSYNC on, it goes as follows:
client sends data -> chunkserver receives data -> chunkserver writes data to disk -> chunkservers syncs data so it's flushed from cache to hardware -> client receives information that operation succeeded

This seems to be ok.

Race condition probably is not the right term, I was referring to a failure between `write` and `fsync`, but as long chunkserver sends ack only after `fsync` there is no issue.

But, in this way would be impossible for a client to write data in asynchronous mode, if they need.
For example, some not-so-important data could be wrote in asynchronous.

**Comment by guestisp** [comment]
> But, in this way would be impossible for a client to write data in asynchronous mode, if they need.
> For example, some not-so-important data could be wrote in asynchronous.

I think there is a way: asking client to ignore the ACK returned from the chunkserver, in example, by using a simple "dd" without `oflag=direct'. In this way, Lizard still returns the ack to the client after a successfull write, but client is ignoring that by still writing even if the previous data wasn't acknoledged.

**Comment by Blackpaw** [comment]
> But, in this way would be impossible for a client to write data in asynchronous mode, if they need.
> For example, some not-so-important data could be wrote in asynchronous.

That would be handled either at the fuse layer - file is written async and fuse returns without waiting for an ack, or the VM layer where you would use qemu writeback mode

**Comment by guestisp** [comment]
> That would be handled either at the fuse layer - file is written async and fuse returns without waiting for an ack, or the VM layer where you would use qemu writeback mode

Right, but i'm still observing very strange speed issue. Today, i'm unable to go faster than 100-110MB/s (more or less a single disk write speed) where yesterday was able to reach 180MB/S

But inside a KVM VM (proxmox with cache=none https://pve.proxmox.com/wiki/Performance_Tweaks#Small_Overview ), it's impossible to go more than 40MB/s

**Comment by guestisp** [comment]
What i'm trying to figure out is why, inside a VM, with:
`dd if=/dev/zero of=test bs=1M count=5000 oflag=sync` i'm getting about 30-40MB/s (and stracing a chunkserver process I can see tons of `open`, `fsync` calls) and why with
`dd if=/dev/zero of=test bs=1M count=5000`
i'm getting 1.GB/s (that's why i'm not using any direct I/O) but the same strace doesn't show any `open` or `fsync` running longer than "dd"

I would expect that "dd" returns more or less immediatly even if chunkserver is still writing in background.

**Comment by pbeza** [comment]
@guestisp Have you found what was the cause of the difference in write speed between writing from the host to the mfs mount point (~180 MB/s) and writing from a VM booted from LizardFS (~30 MB/s)?
